node/src/components/linear_chain_sync.rs

```diff
@@ -265,40 +285,60 @@ impl<I: Clone + PartialEq + 'static> LinearChainSync<I> {
         // Reset peers before creating new requests.
         self.reset_peers(rng);
         let block_height = block_header.height();
-        let curr_state = mem::replace(&mut self.state, State::None);
+        let mut curr_state = mem::replace(&mut self.state, State::None);
         match curr_state {
             State::None | State::Done => panic!("Block handled when in {:?} state.", &curr_state),
+            // Keep syncing from genesis if we haven't reached the trusted block hash
+            State::SyncingTrustedHash {
+                highest_block_seen,
+                ref mut validator_weights,
+                ..
+            } if highest_block_seen != block_height => {
+                if let Some(validator_weights_for_new_era) =
+                    block_header.next_era_validator_weights()
+                {
+                    *validator_weights = validator_weights_for_new_era.clone();
+                }
+                self.state = curr_state;
+                self.fetch_next_block_deploys(effect_builder)
+            }
+            // Otherwise transition to State::SyncingDescendants
```

Comment:
I think the approach with conditions on match patterns is dangerous as it's error-prone. If this and the next arms were switched places then the logic changes. That's why I put it under one match case. It was deliberate.